File: 547-number-of-provinces/number-of-provinces.cpp
class Solution {
public:
    void dfs(int idx, vector<int>& vis, vector<vector<int>>& adj){
        vis[idx] = 1;
        for(int i = 0;i<adj[idx-1].size();i++){
            if(adj[idx-1][i] == 1 && vis[i+1] == 0){
                dfs(i+1,vis,adj);
            }
        }
    }
    int findCircleNum(vector<vector<int>>& adj) {
        int n = adj.size();
        int cnt = 0;
        vector<int> vis(n+1,0);
        for(int i = 1;i<=n;i++){
            if(vis[i] == 0){
                cnt++;
                dfs(i,vis,adj);
            }
        }
        return cnt;
    }
};

File: 547-number-of-provinces/number-of-provinces.java
class Solution {
    public void dfs(int idx, int[] vis, int[][] arr){
        vis[idx] = 1;
        for(int i = 0;i<arr[idx].length;i++){
            if(arr[idx][i] == 1 && vis[i] == -1){
                dfs(i,vis,arr);
            }
        }
    }
    public int findCircleNum(int[][] arr) {
        int n = arr.length;
        int vis[] = new int[n];
        Arrays.fill(vis,-1);
        int cnt = 0;
        for(int i = 0;i<n;i++){
            if(vis[i] == -1){
                dfs(i,vis,arr);
                cnt++;
            }
        }
        return cnt;
    }
}